File: elixir/lib/day01.ex
defmodule Day01 do
  defp convert_input_to_numbers(input) when is_binary(input) do
    input
    |> String.split(",")
    |> Enum.map(&String.trim/1)
    |> Enum.filter(fn x -> String.length(x) > 0 end)
    |> Enum.map(&String.to_integer/1)
  end

  def calculate(input) when is_binary(input) do
    input
    |> convert_input_to_numbers()
    |> Enum.reduce(fn x, acc -> x + acc end)
  end

  def find_reached_twice(input) when is_binary(input) do
    input
    |> convert_input_to_numbers()
    |> _find_reached_twice()
  end

  defp _find_reached_twice(values, frequency \\ 0, counter \\ %{}, found \\ false) do
    if found do
      frequency
    else
      [head | tail] = values

      if Map.has_key?(counter, frequency) do
        _find_reached_twice(tail ++ [head], frequency, counter, true)
      else
        _find_reached_twice(
          tail ++ [head],
          frequency + head,
          Map.put(counter, frequency, 1),
          false
        )
      end
    end
  end

  def run() do
    raw =
      File.read!("../inputs/day01.txt")
      |> String.replace("\n", ",")

    result =
      raw
      |> calculate()

    reached_twice =
      raw
      |> find_reached_twice()

    IO.puts("The reached frecuency is #{result}")
    IO.puts("The first frequency reached twice is #{reached_twice}")
  end
end

File: python/day01.py
from collections import Counter
from itertools import cycle


def calculate(input_):
    """
    Given a string with numbers separated by comma like "+1, -2"
    calculate the sum of the numbers
    """
    return sum([int(x.strip()) for x in input_.split(",") if x])


def find_reached_twice(input_):
    """
    Given a string with numbers separated by comma like "+1, -2"
    calculate the frequency reached twice
    """
    counter = Counter()
    values = cycle([int(x.strip()) for x in input_.split(",") if x])
    frequency = 0
    counter[frequency] += 1
    for value in values:
        frequency += value
        counter[frequency] += 1
        if counter[frequency] == 2:
            break
    return frequency


def run():
    with open("../inputs/day01.txt") as f:
        raw = f.read().replace("\n", ",")
        print(f"The reached frequency is {calculate(raw)}")
        print(f"The first frequency reached twice is {find_reached_twice(raw)}")
